src/module_3/push_exploration.md

```diff
@@ -0,0 +1,5364 @@
+
+
+
+
+```python
+precision, recall, thresholds = precision_recall_curve(y_val, y_val_probs)
+
+desired_recall = 0.32
+closest_recall_idx = np.argmin(np.abs(recall - desired_recall))
+
+# Get the threshold corresponding to that index
+threshold_at_desired_recall = thresholds[closest_recall_idx]
+print(f"Threshold for Recall = {desired_recall}: {threshold_at_desired_recall}")
+```
+
+    Threshold for Recall = 0.32: 0.8416240756662322
+
+
+
+```python
+custom_threshold = 0.84
+
+# Calculate probabilities for the positive class
+y_val_probs = model.predict_proba(X_val)[:, 1]
+y_val_custom_threshold = (y_val_probs > custom_threshold).astype(int)
+
+
+print("Classification Report:")
+print(classification_report(y_val, y_val_custom_threshold))
+
+print("Confusion Matrix:")
+print(confusion_matrix(y_val, y_val_custom_threshold))
+```
+
+    Classification Report:
+                  precision    recall  f1-score   support
+    
+               0       0.99      0.98      0.98    213182
+               1       0.18      0.32      0.23      3213
+    
+        accuracy                           0.97    216395
+       macro avg       0.59      0.65      0.61    216395
+    weighted avg       0.98      0.97      0.97    216395
+    
+    Confusion Matrix:
+    [[208529   4653]
+     [  2180   1033]]
+
+
+We see we misscalculated 2k positive values as negative (false negative), and we generated 4k false positives. However, we were right for 1k purchases realised. The precision for purchase was 18%, which compared to our current open rate, seems OK.
+
+
+```python
+custom_threshold = 0.7
+
+# Calculate probabilities for the positive class
+y_val_probs = model.predict_proba(X_val)[:, 1]
+y_val_custom_threshold = (y_val_probs > custom_threshold).astype(int)
+
+
+print("Classification Report:")
+print(classification_report(y_val, y_val_custom_threshold))
+
+print("Confusion Matrix:")
+print(confusion_matrix(y_val, y_val_custom_threshold))
+```
+
+    Classification Report:
+                  precision    recall  f1-score   support
+    
+               0       0.99      0.97      0.98    213182
+               1       0.14      0.35      0.20      3213
+    
+        accuracy                           0.96    216395
+       macro avg       0.57      0.66      0.59    216395
+    weighted avg       0.98      0.96      0.97    216395
+    
+    Confusion Matrix:
+    [[206260   6922]
+     [  2084   1129]]
+
+
+Now there is the trade-off betweem how many we want to capture and generate as false positive vs the total we are able to dismiss, so the max number for false negatives. As per the problem's statement, we would like to avoid false positives as they might churn – hence lower precision for positives but also lower recall. In this sense, we might select P = 0.84 instead of 0.7
```

Comment:
> we might select P = 0.84 instead of 0.7
here you are talking about thresholds, right? why are you calling it `P`?

Reply:
I think the business reasoning makes sense so you just need to create a metric to reflect it: put a price to the false positives, put a price to the false negatives and see how many you lose / gain for each different threshold